File: src/main/java/fr/raconteur/simpleskinswapper/gui/SkinEntry.java
package fr.raconteur.simpleskinswapper.gui;

import fr.raconteur.simpleskinswapper.SimpleSkinSwapper;
import net.fabricmc.loader.api.FabricLoader;
import net.minecraft.resources.Identifier;
import java.io.File;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

/**
 * Represents a skin file entry in the carousel.
 * Lazily loads the GPU texture on first render.
 */
public class SkinEntry {

    public final File file;
    public final String displayName;
    public SkinType skinType;

    /** GPU texture identifier, null until loaded. */
    public Identifier textureId = null;
    public boolean textureLoading = false;

    public SkinEntry(File file) {
        this.file = file;
        SkinType detected = SkinUtils.detectSkinType(file);
        this.skinType = SkinTypeStore.getType(file.getName(), detected);

        // Display name: filename without extension
        String name = file.getName();
        this.displayName = name.endsWith(".png") ? name.substring(0, name.length() - 4) : name;
    }

    /**
     * Scan the game's skins/ directory and return all .png files as SkinEntry objects.
     */
    public static List<SkinEntry> loadSkins() {
        Path skinsDir = FabricLoader.getInstance().getGameDir().resolve("skins");
        File dir = skinsDir.toFile();
        if (!dir.exists()) {
            dir.mkdirs();
        }
        List<SkinEntry> entries = new ArrayList<>();
        File[] files = dir.listFiles((d, name) -> name.toLowerCase().endsWith(".png"));
        if (files != null) {
            for (File file : files) {
                entries.add(new SkinEntry(file));
            }
        }
        return entries;
    }

    /**
     * Ensure the GPU texture is loaded. Call from render thread.
     * No-op if already loading or loaded.
     */
    public void ensureTextureLoaded() {
        if (textureId != null || textureLoading) return;
        textureLoading = true;

        String sanitized = file.getName()
                .toLowerCase()
                .replaceAll("[^a-z0-9_/.-]", "_");
        String pathHash = String.format("%08x", file.getAbsolutePath().hashCode() & 0x7FFFFFFF);
        String key = "skin/entry_" + sanitized + "_" + pathHash;

        SkinUtils.loadSkinTextureAsync(file, key, id -> {
            this.textureId = id;
            SimpleSkinSwapper.LOGGER.debug("Loaded skin entry texture: {}", file.getName());
        });
    }
}

File: src/main/kotlin/fr/raconteur/simpleskinswapper/gui/SkinEntry.kt
package fr.raconteur.simpleskinswapper.gui

import fr.raconteur.simpleskinswapper.SimpleSkinSwapper
import fr.raconteur.simpleskinswapper.library.SkinRecord
import net.fabricmc.loader.api.FabricLoader
import net.minecraft.resources.Identifier
import java.io.File

/**
 * Represents a skin file entry in the carousel.
 * Lazily loads the GPU texture on first render.
 */
class SkinEntry(@JvmField var file: File) {

    /** File name without extension, used when no display name override is set. */
    @JvmField
    val baseName: String

    /** Optional user-set display name (null = show the file name). */
    @JvmField
    var displayNameOverride: String? = SkinNames.getName(file.name)

    /** Name shown in the UI: the override when set, the file name otherwise. */
    val displayName: String
        get() = displayNameOverride ?: baseName

    @JvmField
    var skinType: SkinType

    /** Registry id of the skin this entry renders (empty for legacy scan entries). */
    @JvmField
    var skinId: String = ""

    /** GPU texture identifier, null until loaded. */
    @JvmField
    var textureId: Identifier? = null

    @JvmField
    var textureLoading = false

    init {
        val detected = SkinUtils.detectSkinType(file)
        skinType = SkinTypes.getType(file.name, detected)

        // Base display name: filename without extension
        val name = file.name
        baseName = if (name.endsWith(".png")) name.substring(0, name.length - 4) else name
    }

    /**
     * Ensure the GPU texture is loaded. Call from render thread.
     * No-op if already loading or loaded.
     */
    fun ensureTextureLoaded() {
        if (textureId != null || textureLoading) return
        textureLoading = true

        val sanitized = file.name
            .lowercase()
            .replace(Regex("[^a-z0-9_/.-]"), "_")
        val pathHash = "%08x".format(file.absolutePath.hashCode() and 0x7FFFFFFF)
        val key = "skin/entry_${sanitized}_$pathHash"

        SkinUtils.loadSkinTextureAsync(file, key) { id ->
            this.textureId = id
            SimpleSkinSwapper.LOGGER.debug("Loaded skin entry texture: {}", file.name)
        }
    }

    companion object {
        /** Builds the render entry for a registry record (registry-backed views). */
        @JvmStatic
        fun fromRecord(record: SkinRecord): SkinEntry {
            val file = FabricLoader.getInstance().gameDir.resolve("skins").resolve(record.file).toFile()
            val entry = SkinEntry(file)
            entry.skinId = record.id
            if (record.name.isNotBlank()) entry.displayNameOverride = record.name
            entry.skinType = if (record.model == SkinRecord.MODEL_SLIM) SkinType.SLIM else SkinType.CLASSIC
            return entry
        }

        /**
         * Scan the game's skins/ directory and return all .png files as SkinEntry objects.
         */
        @JvmStatic
        fun loadSkins(): List<SkinEntry> {
            val skinsDir = FabricLoader.getInstance().gameDir.resolve("skins")
            val dir = skinsDir.toFile()
            if (!dir.exists()) {
                dir.mkdirs()
            }
            val entries = ArrayList<SkinEntry>()
            val files = dir.listFiles { _, name -> name.lowercase().endsWith(".png") }
            if (files != null) {
                for (file in files) {
                    entries.add(SkinEntry(file))
                }
            }
            return entries
        }
    }
}
